src/pwm.rs

```diff
@@ -209,6 +234,93 @@ pub struct C2;
 pub struct C3;
 pub struct C4;
 
+/*
+  The following implemention of the embedded_hal::Pwm uses Hertz as a time type.  This was choosen 
+  because of the timescales of operationsbeing on the order of nanoseconds and not being able to 
+  efficently represent a float on the hardware.  It might be possible to change the time type to
+  a different time based using such as the nanosecond.  The issue with doing so is that the max 
+  delay would then be at just a little over 2 seconds because of the 32 bit depth of the number.
+  Using milliseconds is also an option, however, using this as a base unit means that only there 
+  could be resolution issues when trying to get a specific value, because of the integer nature.
+
+  To find a middle ground, the Hertz type is used as a base here and the Into trait has been 
+  defined for several base time units.  This will allow for calling the set_period method with 
+  something that is natural to both the MCU and the end user.
+*/
+macro_rules! pwm_impl {
+    ( $( $TIMX:ident, $timX:ident, ( $($CHNUM:tt),+ ), ( $($ENCHX:ident),+ ); )+ )  => {
+            $(
+            #[allow(unused_parens)]
+            impl hal::Pwm for Pwm<$TIMX, ($(PwmChannel<$TIMX, $ENCHX>),+,)> {
+                type Channel = u8;
+                type Duty = u16;
+                type Time = Hertz;
+
+                fn enable(&mut self, channel: Self::Channel) {
+                    match channel {
+                        $( $CHNUM => self.$CHNUM.enable(),)+
+                        _ => {}
+                    }
+                }
+
+                fn disable(&mut self, channel: Self::Channel) {
+                    match channel {
+                        $( $CHNUM => self.$CHNUM.disable(),)+
+                        _ => {}
+                    }
+                }
+
+                fn get_duty(&self, channel: Self::Channel) -> Self::Duty {
+                    match channel {
+                        $( $CHNUM => self.$CHNUM.get_duty(),)+
+                        _ => 0
+                    }
+                }
+
+                fn set_duty(&mut self, channel: Self::Channel, duty: Self::Duty) {
+                    match channel {
+                        $( $CHNUM => self.$CHNUM.set_duty(duty),)+
+                        _ => {}
+                    }
+                }
+
+                fn get_max_duty(&self) -> Self::Duty {
+                    unsafe { (*$TIMX::ptr()).arr.read().arr().bits() }
+                }
+
+                fn get_period(&self) -> Self::Time {
+                    let clk = self.clk;
+                    let mut psc: u16 = 0;
+                    let mut arr: u16 = 0;
+                    unsafe {
+                        psc = (*$TIMX::ptr()).psc.read().psc().bits();
+                        arr = (*$TIMX::ptr()).arr.read().arr().bits();
+                    }
```

Comment:
I prefer 
```rust
let mut psc = unsafe{*$TIMX::...};
```

Reply:
Ohh, I like that better as well.  Again, will do this tonight.

Reply:
Corrected